tests/net/adversary.rs

```diff
@@ -435,9 +525,83 @@ where
 {
     #[inline]
     fn pre_crank<R: Rng>(&mut self, mut net: NetMutHandle<'_, D, Self>, rng: &mut R) {
-        let l = net.0.messages_len();
-        if l > 0 {
-            net.swap_messages(0, rng.gen_range(0, l));
+        swap_random(&mut net, rng);
+    }
+}
+
+/// An adversary that performs naive replay attacks.
+///
+/// The adversary will randomly take a message that is sent to one of its nodes
+/// and re-send it to a different node.
+///
+/// Additionally it will broadcast unrelated random messages,
+/// and the message queue by a random node id.
+#[derive(Copy, Clone, Debug, Default)]
+pub struct RandomAdversary {
+    /// Probability of a message replay
+    p_replay: f32,
+    /// Probability of a message injection
+    p_inject: f32,
+}
+
+impl RandomAdversary {
+    pub fn new(p_replay: f32, p_inject: f32) -> Self {
+        RandomAdversary { p_replay, p_inject }
+    }
+}
+
+impl<D> Adversary<D> for RandomAdversary
+where
+    D: ConsensusProtocol,
+    D::Message: Clone,
+    D::Output: Clone,
+    rand::distributions::Standard:
+        rand::distributions::Distribution<<D as ConsensusProtocol>::Message>,
+{
+    #[inline]
+    fn pre_crank<R: Rng>(&mut self, mut net: NetMutHandle<'_, D, Self>, rng: &mut R) {
+        sort_by_random_node(&mut net, rng);
+    }
+
+    #[inline]
+    fn tamper<R: Rng>(
+        &mut self,
+        mut net: NetMutHandle<'_, D, Self>,
+        msg: NetMessage<D>,
+        rng: &mut R,
+    ) -> Result<CpStep<D>, CrankError<D>> {
+        // only replay a message in some cases
+        if randomly(self.p_replay) {
```

Comment:
The old `randomly` doesn't fit in the `net` framework because it leads to non-reproducible tests due to the use of `thread_rng`. Can you use `rng` here instead?

Reply:
Switched to TestRng